server.go

```diff
@@ -4173,13 +4327,297 @@ func (s *server) SubscribeCustomMessages() (*subscribe.Client, error) {
+
+		// A restricted-access slot has opened up.
+		s.numRestricted -= 1
+
+		s.peerScores[remoteHex] = peerSlotStatus{state: temporary}
+
+	default:
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
+	}
+
+	return nil
+}
+
+// NewPendingCloseChan is called when a pending-open channel prematurely closes
+// before the funding transaction has confirmed. This potentially demotes a
+// temporary-access peer to a restricted-access peer. If no restricted-access
+// slots are available, the peer will be disconnected.
+func (s *server) NewPendingCloseChan(remoteHex string) error {
+	// Ensure that we don't lock banScoreMtx and _then_ lock the server
+	// mutex. This could lead to deadlock.
+	s.banScoreMtx.Lock()
+
+	// Fetch the peer's access status from peerScores.
+	status, found := s.peerScores[remoteHex]
+	if !found {
+		s.banScoreMtx.Unlock()
+
+		// If we didn't find the peer, we'll return an error.
+		return ErrNoPeerScore
+	}
+
+	switch status.state {
+	case protected:
+		s.banScoreMtx.Unlock()
+
+		// If this peer is protected, we don't do anything.
+		return nil
+
+	case temporary:
+		// If this peer is temporary, we need to check if it will revert
+		// to a restricted-access peer.
+		numPending, pendingFound := s.pendingChannelPeers[remoteHex]
+		if !pendingFound {
+			s.banScoreMtx.Unlock()
+
+			// Error if we did not find any info in
+			// pendingChannelPeers.
+			return ErrNoPendingPeerInfo
+		}
+
+		currentNumPending := numPending - 1
+
+		if currentNumPending == 0 {
+			// Remove the entry from pendingChannelPeers.
+			delete(s.pendingChannelPeers, remoteHex)
+
+			// If this is the only pending-open channel for this
+			// peer and it's getting removed, attempt to demote
+			// this peer to a restricted peer.
+			if s.numRestricted == maxRestrictedSlots {
+				// There are no available restricted slots, so
+				// we need to disconnect this peer.
+				s.banScoreMtx.Unlock()
+
+				remotePub, err := btcec.ParsePubKey(
+					[]byte(remoteHex),
+				)
+				if err != nil {
+					return err
+				}
+
+				// The banScoreMtx cannot be held when the
+				// below DisconnectPeer call occurs.
+				return s.DisconnectPeer(remotePub)
+			}
+
+			// Otherwise, there is an available restricted-access
+			// slot, so we can demote this peer.
+			s.peerScores[remoteHex] = peerSlotStatus{
+				state: restricted,
+			}
+
+			// Update numRestricted.
+			s.numRestricted++
+
+			s.banScoreMtx.Unlock()
+
+			return nil
+		}
+
+		// Else, we don't need to demote this peer since it has other
+		// pending-open channels with us.
+		s.pendingChannelPeers[remoteHex] = currentNumPending
+
+		s.banScoreMtx.Unlock()
+
+		return nil
+
+	case restricted:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible. This indicates an error.
+		return fmt.Errorf("invalid peer access state transition")
+
+	default:
+		s.banScoreMtx.Unlock()
+
+		// This should not be possible.
+		return fmt.Errorf("invalid peer access status")
```

Comment:
nit: should we add the status to the error?

Reply:
It's unreachable code, only added for completeness